File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <unordered_set>

using namespace std;

vector<int> solution(int n, vector<string> words) {
    unordered_set<string> used;
    used.insert(words[0]);
    
    for (int i = 1; i < words.size(); ++i) {
        if (!used.insert(words[i]).second || words[i].front() != words[i-1].back()) {
            return vector<int>{i % n + 1, i / n + 1};
        }
    }
    
    return vector<int>{0, 0};
}

File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.java
import java.util.Set;
import java.util.HashSet;

class Solution {
    public int[] solution(int n, String[] words) {
        Set<String> used = new HashSet<>();
        char expectedStart = words[0].charAt(0);

        for (int i = 0; i < words.length; i++) {
            String word = words[i];

            // 1) 시작 글자 검사
            if (word.charAt(0) != expectedStart) {
                return new int[] {(i % n) + 1, (i / n) + 1};
            }
            
            // 2) 중복 검사
            if (!used.add(word)) {
                return new int[] {(i % n) + 1, (i / n) + 1};
            }

            // 다음 예상 시작 글자 갱신
            expectedStart = word.charAt(word.length() - 1);
        }

        return new int[] {0, 0};
    }
}